incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,174 @@
+        # these are the metacoalitions of size <= limit_of_revealed
+        self.viable_metacoalitions = len(self.meta_rank_to_id)
+        # map metacoalition id -> metacoalition size rank
+        self.meta_id_to_rank = np.zeros(self.viable_metacoalitions, dtype=int)
+        self.meta_id_to_rank[self.meta_rank_to_id] = np.arange(self.viable_metacoalitions)
+
+        # regret minimizer for every internal node -- except for the bottom layer of game tree
+        self.number_of_regret_minimizers = coalitions_up_to(self.number_of_coalitions, limit_of_revealed - 1)
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.cumulative_regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+        self.cumulative_strategy = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
+                                            dtype=RMValue)
+        self.iteration = 0
+
+    def get_metacoalition_id(self, metacoalition: Iterable[Coalition]) -> int:
+        """Properly get an id of a metacoalition."""
+        coalition_player_ids = self.coalitions_to_player_ids[
+            [x.id for x in metacoalition if len(x) not in [0, 1, 2**self.number_of_players]]]
+        return np.sum(2 ** coalition_player_ids)
+
+    def regret_matching_strategy(self, past_actions: Iterable[Coalition] | int) -> np.ndarray:
+        """Get the action predicted by the strategy using the Regret Matching minimizer.
+
+        If past_actions is an int, it is considered a metacoalition id.
+        Otherwise, it is considered as a list of coalitions of the original game.
+
+        This returns the current prediction, not the averaged strategy over time.
+        """
+        metacoalition = past_actions if isinstance(past_actions, int) else self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        # TODO: predictive: přičte predikci ...
+        positive_regret = self.cumulative_regret[rank] * (self.cumulative_regret[rank] > 0)
+        if positive_regret.sum() == 0:
+            positive_regret = np.ones(self.number_of_coalitions)
+            used_coalitions = list(Coalition(metacoalition).players)
+            positive_regret[used_coalitions] = 0
+        return positive_regret / positive_regret.sum()
+
+    def get_average_strategy(self, past_actions: Iterable[Coalition]) -> np.ndarray:
+        """Get average strategy, in terms of actual coalitions in the original game."""
+        metacoalition = self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        # in `regret minimization` we only add the "previous" regret matching strategy, not the current one,
+        # so we need to add it here
+        average_strategy = self.cumulative_strategy[rank] + self.regret_matching_strategy(metacoalition)
+        average_strategy = average_strategy / average_strategy.sum()
```

Comment:
is this correct?

Reply:
I don't see why you need to add the `regret_matching_strategy` here. It probably depends on when you call this, but I would expect this method to give me just the normalized `self.cumulative_strategy`.

Reply:
Here's how I imagine it:
1. We "train" the minimizer, essentially repeatedly calling the `regret_min_iteration` method. This always updates the cumulative losses and the cumulative strategy, and _then_ it updates the cumulative regret.
2. We ask for the current average strategy. But the current cumulative regret wasn't yet used to update the cumulative strategy (because in `regret_min_iteration`, it was updated _after_ the cumulative strategy).

So basically, I feel like if we don't add the regret matching strategy, we have one "update of regret" we're not using.
This is my intuition, is it correct? Also, is that how it is done in the literature?

Reply:
I think it is technically correct, but I really don't like this design. For example, we can later add a second function which needs the `average_strategy`, and this extra evaluation will ruin the convergence. Accumulating regret should be done in the main loop.

Reply:
Ok. Changed.